plugins/module_utils/network/sonic/utils/formatted_diff_utils.py

```diff
@@ -0,0 +1,427 @@
+#
+# -*- coding: utf-8 -*-
+# Copyright 2023 Dell Inc. or its subsidiaries. All Rights Reserved
+# GNU General Public License v3.0+
+# (see COPYING or https://www.gnu.org/licenses/gpl-3.0.txt)
+
+from __future__ import absolute_import, division, print_function
+
+__metaclass__ = type
+
+from copy import (
+    deepcopy
+)
+from pprint import (
+    pformat
+)
+from difflib import (
+    context_diff
+)
+
+
+def get_key_sets(dict_conf):
+    key_set = set(dict_conf.keys())
+    trival_key_set = set()
+    dict_list_key_set = set()
+    for key in key_set:
+        if dict_conf[key] not in [None, [], {}]:
+            if isinstance(dict_conf[key], (list, dict)):
+                dict_list_key_set.add(key)
+            else:
+                trival_key_set.add(key)
+    return trival_key_set, dict_list_key_set
+
+#
+# Pre-defined Delete Operations
+#
+
+
+def __DELETE_CONFIG(key_set, command, exist_conf):
+    new_conf = []
+    return True, new_conf
+
+
+def __DELETE_CONFIG_IF_NO_SUBCONFIG(key_set, command, exist_conf):
+    nu, dict_list_cmd_key_set = get_key_sets(command)
+    if len(dict_list_cmd_key_set) == 0:
+        new_conf = []
+        return True, new_conf
+    else:
+        new_conf = exist_conf
+        return False, new_conf
+
+
+def __DELETE_SUBCONFIG_AND_LEAFS(key_set, command, exist_conf):
+    new_conf = exist_conf
+
+    trival_cmd_key_set, dict_list_cmd_key_set = get_key_sets(command)
+    trival_cmd_key_not_key_set = trival_cmd_key_set.difference(key_set)
+    for key in trival_cmd_key_not_key_set:
+        new_conf.pop(key, None)
+
+    nu, dict_list_exist_key_set = get_key_sets(exist_conf)
+    common_dict_list_key_set = dict_list_cmd_key_set.intersection(dict_list_exist_key_set)
+    if len(common_dict_list_key_set) != 0:
+        for key in common_dict_list_key_set:
+            new_conf.pop(key, None)
+
+    return True, new_conf
+
+
+def __DELETE_SUBCONFIG_ONLY(key_set, command, exist_conf):
+    new_conf = exist_conf
+    nu, dict_list_cmd_key_set = get_key_sets(command)
+    nu, dict_list_exist_key_set = get_key_sets(exist_conf)
+    common_dict_list_key_set = dict_list_cmd_key_set.intersection(dict_list_exist_key_set)
+    for key in common_dict_list_key_set:
+        new_conf.pop(key, None)
+    return True, new_conf
+
+
+def __DELETE_OP_DEFAULT(key_set, command, exist_conf):
+    new_conf = exist_conf
+    trival_cmd_key_set, dict_list_cmd_key_set = get_key_sets(command)
+
+    if (len(key_set) == len(trival_cmd_key_set)) and \
+       (len(dict_list_cmd_key_set) == 0):
+        new_conf = []
+        return True, new_conf
+
+    trival_cmd_key_not_key_set = trival_cmd_key_set.difference(key_set)
```

Comment:
If the top level configuration includes "trivial" attributes and also non-trivial attributes, it looks like this will delete all "trivial" attributes with keys in the playbook command regardless of whether the value in the playbook matches the current configured value (because the input "key_set" value will be "None".)

This is because the first recursion, seeing that the command and existing configs are dicts, will end up setting the “key_set” argument to NONE for the second recursion, causing the call to the (default) delete op (__DELETE_OP_DEFAULT) to receive an empty key set, while the trivial key set is not empty. The length mismatch between the empty key_set and the non-empty trivial_cmd_key_set would then cause this function to fall through to the popping of all trivial keys (because the difference subtracts nothing).